src/core/Akka/IO/TcpConnection.cs

```diff
@@ -778,110 +818,191 @@ public WriteFileFailed(SocketException cause)
-
-                        if (_connection.traceLogging)
-                            _connection.Log.Debug("Wrote [{0}] bytes to channel", bytesWritten);
-                        if (bytesWritten < data.Count)
-                        {
-                            // we weren't able to write all bytes from the buffer, so we need to try again later
-                            data = data.Slice(bytesWritten);
-                        }
-                        else // finished writing
-                        {
-                            if(Ack != NoAck.Instance) Commander.Tell(Ack);
-                            Release();
-                            return _connection.CreatePendingWrite(Commander, _tail, info);
-                        }
-                    }
-
+                    _sendArgs.SetBuffer(_dataToSend);
+                    if (!_connection.Socket.SendAsync(_sendArgs))
+                        _self.Tell(SocketSent.Instance);
                 }
                 catch (SocketException e)
                 {
                     _connection.HandleError(info.Handler, e);
-                    return this;
                 }
             }
+        }
 
-            private int Send(ByteString data)
+        public class PendingSimpleWritesQueue
+        {
+            private readonly Option<int> _maxQueueSizeInBytes;
+            private readonly Queue<(SimpleWriteCommand Command, IActorRef Commander, int Size)> _queue;
+            private int _totalSizeInBytes = 0;
+
+            public PendingSimpleWritesQueue(Option<int> maxQueueSizeInBytes)
             {
-                try
+                _maxQueueSizeInBytes = maxQueueSizeInBytes;
+                _queue = new Queue<(SimpleWriteCommand Command, IActorRef Commander, int Size)>();
+            }
+
+            /// <summary>
+            /// Gets total number of items in queue
+            /// </summary>
+            public int ItemsCount => _queue.Count;
+
+            /// <summary>
+            /// Adds all <see cref="SimpleWriteCommand"/> subcommands stored in provided command.
+            /// Performs buffer size checks
+            /// </summary>
+            /// <exception cref="InternalBufferOverflowException">
+            /// Thrown when data to buffer is larger then allowed <see cref="_maxQueueSizeInBytes"/>
+            /// </exception>
+            public void EnqueueSimpleWrites(WriteCommand command, IActorRef sender)
+            {
+                foreach (var writeInfo in ExtractFromCommand(command))
                 {
-                    return _connection.Socket.Send(data.Buffers);
+                    var sizeAfterAppending = _totalSizeInBytes + writeInfo.DataSize;
+                    if (_maxQueueSizeInBytes.HasValue && _maxQueueSizeInBytes.Value < sizeAfterAppending)
+                    {
+                        throw new InternalBufferOverflowException($"Could not receive write command of size {writeInfo.DataSize} bytes, " +
```

Comment:
If we're going to reject a write because of the buffer being full, we might be better off just having this method return a `bool` - returns `true` if the write can fit, `false` if it can't. Throwing exceptions is expensive (from a throughput standpoint) and I think this is an area where we can avoid doing it. What do you think @IgorFedchenko @Horusiath ?

Reply:
Indeed, you are right. Will refactor this right now.